neps/nep-0393.md

```diff
@@ -0,0 +1,480 @@
+    /// Must emit `Renew` event.
+    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64);
+
+    /// Revokes SBT, could potentially burn it or update the expire time.
+    /// Must be called by an SBT contract.
+    /// Must emit one of `Revoke` or `Burn` event.
+    /// Returns true if a token is a valid, active SBT. Otherwise returns false.
+    fn sbt_revoke(&mut self, token: u64) -> bool;
+
+    /// Transfers atomically all SBT tokens from one account to another account.
+    /// The caller must be an SBT holder and the `to` must not be a banned account.
+    /// Must emit `Revoke` event.
+    // #[payable]
+    fn sbt_soul_transfer(&mut self, to: AccountId) -> bool;
+}
+```
+
+SBT smart contracts can implement NFT query interface to make it compatible with NFT tools. Note, we use U64 type rather than U128.
+
+```rust
+trait SBTNFT {
+    fn nft_total_supply(&self) -> U64;
+    // here we index by token id instead of by class id (as done in `sbt_tokens_by_owner`)
+    fn nft_tokens_for_owner(&self, account_id: AccountId, from_index: Option<U64>, limit: Option<u64>) -> Vec<Token>;
+    fn nft_supply_for_owner(&self, account_id: AccountId) -> U64;
+}
+```
+
+### Events
+
+Event design principles:
+
+- Events don't need to repeat all function arguments - these are easy to retrieve by indexer (events are consumed by indexers anyway).
+- Events must include fields necessary to identify subject matters related to use case.
+- When possible, events should contain aggregated data, with respect to the standard function related to the event.
+
+```typescript
+// only valid integer numbers (without rounding errors).
+type u64 = number;
+
+type Nep393Event {
+  standard: "nep393";
+  version: "1.0.0";
+  event: "mint" | "recover" | "renew" | "revoke" | "burn" | "ban" | "soul_transfer" ;
+  data: Mint | Recover | Renew | Revoke | Burn | Ban[] | SoulTransfer;
+}
+
+/// An event minted by the Registry when new SBT is created.
+type Mint {
+  ctr: AccountId;    // SBT Contract recovering the tokens
+  tokens: [[AccountId, u64[]]];  // list of pairs (token owner, TokenId[])
+}
+
+/// An event emitted when a recovery process succeeded to reassign SBT, usually due to account
+/// access loss. This action is usually requested by the owner, but executed by an issuer,
+/// and doesn't trigger Soul Transfer. Registry recovers all tokens assigned to `old_owner`,
+/// hence we don't need to enumerate them.
+/// Must be emitted by an SBT registry.
+type Recover {
+  ctr: AccountId         // SBT Contract recovering the tokens
+  old_owner: AccountId;  // current holder of the SBT
+  new_owner: AccountId;  // destination account.
+}
+
+/// An event emitted when a existing tokens are renewed.
+/// Must be emitted by an SBT registry.
+type Renew {
+  ctr: AccountId;  // SBT Contract renewing the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when an existing tokens are revoked.
+/// Revoked tokens should not be listed in a wallet.
```

Comment:
Can we be specific in the comments and state that the _Revoked tokens will continue to be listed by the registry_.

I assume this is the case here, unlike in the case of a Burn event.

Reply:
yes, thanks for suggestion.